takashi-task-framework.md

```diff
@@ -0,0 +1,265 @@
+```
+
+We can also "jail" each task, by using a FreeBSD-esque jail. Each task
+will be granted its own process space, memory, and disk, and cannot
+escape (in case malicious or fault code is executed, it won't be able to
+cat /etc/passwd or other evil things).
+
+## Logging
+
+We need to make sure that logging is prioritized, since this is
+complicated. If I write a task file and one of the dependencies fails to
+install, this is a problem as well. We'll surface a command like
+`task logs [$NAME]`
+which will let the user query the task runner for the stdout of a task name.
+
+## Running at Scale
+
+Since we can assume that each task is idempotent (we'll jail each task
+and assume they can run in any order, including in parallel), we can
+allocate a cluster of machines to run tasks for us, and there can be a
+centralized metadata service that users can query for the status of the
+jobs they care about, along with audits of status changes, which they
+can query by running `task audit [NAME]`.
+
+We can assume that running 1B tasks per day is about 10k tasks per
+second. We can try to keep a cluster of a thousand machines to run jobs,
+all responding back to a centralized system that is replicated to store
+metadata on tasks. The metadata server can generate ids for each task,
+and also queries the tasks that need to run (cron jobs) and runs the
+dependencies of all tasks provided, to know when to run a task if its
+`RunIf` block wants it to run.
+
+The centralized system could use a backing store of a time-series like
+database, or it could use an append-only system, where it appends each
+event that happens to a table, and a query will be indexed to make sure
+it stays efficient (instead of table scanning).
+
+We can create a status table that contains enums like so:
+(this allows us to add more statuses in the future, just like errno in C).
+
+```sql
+create table status (
+  id primary key auto increment,
+  name varchar(255)
+);
+```
+
+```json
+{ "id": 1, "job_id": 3252, "status": 1, "timestamp": 17281 } # job 3252 completed at time 17281
+{ "id": 2, "job_id": 3253, "status": 2, "timestamp": 17283 } # job 3253
+is in progress
+```
+
+We can then query by `job_id` to see the status of jobs we care for, and
+query with some SQL like syntax.
+
+## MapReduce Jobs
+
+We'll also run into the problem where some tasks must run on multiple
+machines. If tasks can be run on multiple machines, we'll allow the user
+to pass map-reduce like syntax (provide a map function), which the
+centralized system will partition to open machines, and then finally a
+reduce function which the centralized system will aggregate and then
+finally write to a datastore (or a file).
+
+We could allow for the user to specify that a task is a mapreduce job,
+with the caveat that this requires tasks to be completely side effect
+free, and our system must be able to rollback or otherwise kill the job
+if it fails part way through (like a transaction).
```

Comment:
I like the idea of batch processing especially if we pair the batches with individual levels of the task tree. How do we properly estimate these batches? how do we ensure that the bulk inserts dont cause some insert bottleneck though? would a system like this be more write heavy? is there anything we can do to increase the speed of our bulk insertion since we will be doing this a lot?

Reply:
1. How do we properly estimate these batches?

- We'd need some way to calculate this. Luckily, our configuration language has limits on memory/hard disk space, so we could have our centralized service try to find workers to place batch jobs on by greedily finding some large jobs, medium jobs, and small jobs, and place them on the same worker. 

2. how do we ensure that the bulk inserts dont cause some insert bottleneck though?
- If we have a problem with bulk inserts causing a bottleneck, we can batch writes.
- Assume that we run 10k jobs/second, and state is an FSA where jobs transition to created and then to in-progress and then to either success or failure, each job will do 3 writes to the backing data store. We'd roughly have 30k writes per second to the backing data store on average, and maybe assume 3x that for a spike (90k writes per second). Since this is for 1 record, and the database is append only, we can batch writes by doing them something like 10 times a second, for a max write throughput of 9k records 10 times a second. This shouldn't be a big deal for our backing data store to keep track of.
- We can also disable indexes, since we're only doing appends.